### ISSUE TITLE
OPEN-1B: A Fully Auditable Training Run (2026.09)

### ISSUE THREAD

**Issue by AtsunoriFujita** [AtsunoriFujita/paper-reading-notes#188]
# OPEN-1B: A Fully Auditable Training Run

https://arxiv.org/abs/2609.17380

### 問題意識

- **知能の民主化は、LLM へのアクセスと訓練過程の透明性の両方なしには達成できない**。現状の AI は**透明性の3階層**を形成している。
  - **クローズドモデル**（GPT Sol・Claude Opus・Gemini・Grok）: **API でアクセスできてもアーキも「そのモデルに至った軌跡」も知りえない**。**Anthropic と OpenAI が旗艦モデルをゲートキープした例**や、**サブスクリプションで提供 AI を downgrade しうる**点が挙げられる。**少数企業が決める制約が開かれた自由な研究と AI 利用一般を害する**うえ、**安全のためと称しながら説明責任と責任を暗黙にユーザへ移転している**。
  - **オープンウェイトモデル**（Qwen・Kimi・DeepSeek・GLM）: **重みは公開されるが、訓練データと手法の不透明さの度合いは様々**。**データの透明性が部分的なので、ユーザは訓練中に注入されたバイアスや選好を知りえない**。**政治的意見のようなイデオロギー的バイアスは既に分析されているが、選好バイアスの一部はその領域が分析されるまで検出困難、あるいは不可能**。
  - **オープンソースモデル**（OLMo・Apertus・Marin・Pythia）: 重み・データ・レシピを揃えて公開するので**十分なリソースがあれば似たモデルを再現できる**。
- **だがオープンソースでも「完全には再現可能でない」**。ここが本論文の切り込む隙間である。
  - **浮動小数点演算は非結合的なので、まったく同じレシピに従っても僅かに異なる重みに到達する**。したがって**訓練ステップの正しさを検査できない**。
  - **proof-of-learning や proof-of-training-data のような既存のヒューリスティック検証は確率的な検証しか与えられない**。**バックドアは数点のデータで注入できるので、そうした手法は検出に失敗しうる**。
  - よって**再現可能な演算子を通じて訓練しない限り、オープンソースモデルでさえ完全には検証可能でなく、厳密な訓練過程を保証できない**。
- ここから**モデル開示の次の階層**を提案する——**あらゆるデータサンプルへのあらゆる演算が誰にでも検証可能な完全監査可能モデル**。
- **決定性と再現性の区別**も問題意識の一部である。 **決定性は「同じマシン・同じ環境で毎回同じ結果」** で大半の ML フレームワークが提供するが、 **再現性は「異種ハードウェア（プロセッサと GPU、2つの異なる GPU モデル）を跨いでビット単位で同じ結果」** というより強い要件（**ULP 1 以内でなく完全に同一**）。これを **bitwise reproducibility (BR)** と呼ぶ。**決定性は再現性を含意しない**。
- **BR が壊れる2つの原因**: **(1) 加算の非結合性**——`(a⊕b)⊕c ≠ a⊕(b⊕c)`。**ハードウェアは並列に加算し、異なるハードウェアが異なる順序を割り当てる**（GPU は数千レーンに分割してハードウェア固有パターンで結合、プロセッサは端から端まで累算）。**(2) 同一の高レベル演算が異なる機械命令にコンパイルされ、それぞれ異なる丸めをする**。

### 本論文のアプローチ

- **RepOps（§2.1）**: Gensyn の再現可能演算ライブラリ。**autograd 拡張つきの馴染みやすい API** を提供し、**各演算がプロセッサ・NVIDIA GPU・Apple GPU を跨いで BR な結果を返す**ように書かれている。
- **3層の順序固定（§6）**: BR な監査には**訓練過程全体に確定的な順序を割り当てる**必要があり、対象は **(1) GPU カーネルのリダクション**（線形層の GEMM カーネルや softmax の和リダクション）、**(2) データバッチング**（データ並列クラスタ環境の順序が単一デバイス監査の順序に一致するよう）、**(3) ノード内/ノード間の集団通信**（特に reduce-scatter と all-reduce）。
  - **RepOps がカーネル演算を逐次化**、**topologically invariant datastream（§2.3）がバッチ順序を処理**、**決定論的集団通信を実装**して通信を扱う。
- **監査ハーネス**: これらの逐次化を組み合わせ、**ユーザがクラスタ全体の作業をステップ粒度で再現できる**ようにする。**クラスタの各 GPU の forward/backward を逐次順序で再生し、勾配と集団通信リダクションをローカル累算器で追跡し、オプティマイザを適用して重みを進める**。
- **集合的検証システム**: **コモディティハードウェアでの全訓練実行の検証は非現実的**なので、**オープンソースソフトウェアで特定のモジュールや関数を検証できるのに着想を得て、監査を分散化して集合的に検証する**。**多数のユーザが各ステップを個別に証明することで、訓練実行の厳密な集合的最終状態に到達できる**。
- **決定論的 all-reduce**: NCCL の ring all-reduce は N に対し漸近的に平坦だが、**本論文の決定論的経路は recursive-doubling butterfly を使う**——**リダクション順序を固定する代わりに、`1/N` セグメントでなく `log₂N` ホップの各回でバッファ全体を送る**。

### 評価とAblation Study

**Table 2/3: Open-1B architecture.**

| 項目 | 値 |
|---|---|
| Layers | 24 |
| Hidden Size | 2048 |
| Attention Heads (Q/KV) | 16 / 4 (GQA) |
| Head Dimension | 128 |
| FFN Hidden Size | 5632 |
| Vocabulary Size | 128,256 |
| Sequence Length | 4096 |
| Sliding Window | 512（5層ごとに full） |
| Gradient Clipping | 1.0 |
| LR Warmup | 667 steps |
| LR Schedule (Cosine) | 400B tokens, ピークの 10% まで |
| Total Parameters | **1.61B（非埋め込み 1.08B）** |

**Table 7: Evaluation vs. OLMo 2 1B.**

| Checkpoint | Tokens | Avg | OLMES | MMLU | ARC C | HSwag | WinoG | NQ | DROP | AGIEval | GSM8K | MMLU PRO | TQA |
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
| OLMo 2 1B | 4T | **31.9** | **61.5** | **26.9** | 26.1 | **67.5** | **67.8** | **16.1** | **25.1** | **24.5** | **3.3** | 11.1 | **50.1** |
| **Open-1B** | **400B** | 25.4 | 50.1 | 25.9 | **36.2** | 48.6 | 54.1 | 8.1 | 16.8 | 20.9 | 1.5 | **11.7** | 30.0 |

- **性能では OLMo 2 1B に届かない**。平均 **25.4 vs 31.9**。ただし **Open-1B は 400B トークン、OLMo 2 は 4T トークン**と**10 分の1の訓練量**である点を踏まえるべき。**ARC Challenge（36.2 vs 26.1）と MMLU PRO（11.7 vs 11.1）では上回る**が、**HellaSwag（48.6 vs 67.5）・TQA（30.0 vs 50.1）・NQ（8.1 vs 16.1）では大きく劣る**。
- **再現性のコストが定量化される**（§5）。**GCP の a3-megagpu-8g 6ノードで wall-clock 29.5 日**（うち**実際の事前学習は 27.8 日**、差は障害対応と工学的問題の解決）。**GPU あたり約 485 ワット**（平均 50% 電力、スパイク 70%）。
- **MFU は約 5% で安定**。著者は率直に「**これは最適化された非再現的カーネル集合で達成される MFU より1桁低い。我々は再現性のために計算性能をトレードオフしている**」と述べる。
- **決定論的 all-reduce のオーバーヘッドは2の冪で小さい**。**recursive-doubling butterfly は 2ノードで 1 ホップ、4ノードで 2 ホップなので determinism premium は控えめ**だが、**2の冪でないノード数ではコストが増す**。強スケーリング研究は**global batch を 288 マイクロバッチ（ステップあたり 4,718,592 トークン）に固定して 1〜6 ノードでスケール**させ、**各構成をハッシュあり・なしで2回実行、30 ステップ走らせて 4〜30 を平均**する形で行われた。
- **監査の意義を正確に限定**している。「**ビット単位の精度で各演算を検証できることは、モデルが宣言どおりに訓練されたことを保証する。これはモデルからバイアスを除去しないが、監査者がどのバイアスが存在しうるかを絶対的に確信できるようにする**」。そして「**未開示データでモデルを侵害するバックドア注入攻撃を検出する能力を、初めて手にした**」。
- まとめ: **性能でなく検証可能性を目的とした訓練実行**。**MFU 5%（最適化カーネルの1桁下）という明示的な代償**を払って、**異種ハードウェアを跨ぐビット単位再現性**を達成し、**訓練実行全体を単一ハッシュに帰着させる**。**集合的監査**により、コモディティハードウェアのユーザが分散して各ステップを検証できる設計になっている。

**Comment by AtsunoriFujita** [comment]
https://github.com/gensyn-ai/open-transformers
https://huggingface.co/collections/Gensyn/open-1b
https://open1b.gensyn.ai/
